File: motor3d 2.0/zbuffer.vb
Imports Motor3D.Primitivas3D
Imports Motor3D.Utilidades
Imports Motor3D.Espacio2D
Imports System.Drawing

Namespace Escena
    Public Class ZBuffer
        Inherits ObjetoEscena

        Public Shadows Event Modificado(ByRef Sender As ZBuffer)

        Private mObjetos() As ElementoZBuffer
        Private mRepresentaciones() As Poligono2D
        Private mNumeroIndices As Integer
        Private mTiempo As Double

        Public ReadOnly Property Objetos As ElementoZBuffer()
            Get
                Return mObjetos
            End Get
        End Property

        Public ReadOnly Property NumeroObjetos As Integer
            Get
                If Not mObjetos Is Nothing Then
                    Return mObjetos.GetUpperBound(0) + 1
                Else
                    Return 0
                End If
            End Get
        End Property

        Public ReadOnly Property Represenatciones As Poligono2D()
            Get
                Return mRepresentaciones
            End Get
        End Property

        Public ReadOnly Property Vacio As Boolean
            Get
                Return mObjetos Is Nothing
            End Get
        End Property

        Public ReadOnly Property TiempoUltimoCalculo As Double
            Get
                Return mTiempo
            End Get
        End Property

        Public Sub Actualizar(ByRef Poliedros() As Poliedro, ByRef Camara As Camara3D)
            Dim Objetos As New List(Of ElementoZBuffer)
            Dim Representaciones As New List(Of Poligono2D)
            Dim T As Date = Now
            Dim n As Integer = 0
            If Not Poliedros Is Nothing Then
                If Not Vacio Then
                    For i As Integer = 0 To mObjetos.GetUpperBound(0)
                        If mObjetos(i).Indices(1) <= Poliedros.GetUpperBound(0) AndAlso mObjetos(i).Indices(2) < Poliedros(mObjetos(i).Indices(1)).NumeroCaras Then
                            If Poliedros(mObjetos(i).Indices(1)).CaraVisible(mObjetos(i).Indices(2), Camara) Then
                                Objetos.Add(New ElementoZBuffer(Poliedros(mObjetos(i).Indices(1)).Caras(mObjetos(i).Indices(2)).BaricentroSRC.Z, n, mObjetos(i).Indices(1), mObjetos(i).Indices(2)))
                                Representaciones.Add(New Poligono2D(Poliedros(mObjetos(i).Indices(1)).Caras(mObjetos(i).Indices(2)).Representacion(Poliedros(mObjetos(i).Indices(1)).Vertices)))
                                Representaciones(Representaciones.Count - 1).Color = Poliedros(mObjetos(i).Indices(1)).Caras(mObjetos(i).Indices(2)).Color
                                Poliedros(mObjetos(i).Indices(1)).Caras(mObjetos(i).Indices(2)).CargadaEnBuffer = True

                                n += 1
                            Else
                                Poliedros(mObjetos(i).Indices(1)).Caras(mObjetos(i).Indices(2)).CargadaEnBuffer = False
                            End If
                        End If
                    Next
                End If

                For i As Integer = 0 To Poliedros.GetUpperBound(0)
                    For j As Integer = 0 To Poliedros(i).NumeroCaras - 1
                        If Not Poliedros(i).Caras(j).CargadaEnBuffer Then
                            If Poliedros(i).CaraVisible(j, Camara) Then
                                Objetos.Add(New ElementoZBuffer(Poliedros(i).Caras(j).BaricentroSRC.Z, n, i, j))
                                Representaciones.Add(Poliedros(i).Representacion(j))
                                Poliedros(i).Caras(j).CargadaEnBuffer = True

                                n += 1
                            End If
                        End If

                    Next
                Next

                mObjetos = Objetos.ToArray
                mRepresentaciones = Representaciones.ToArray
                Reordenar()
            Else
                mObjetos = Nothing
                mRepresentaciones = Nothing
            End If

            mTiempo = (Now - T).TotalMilliseconds

            RaiseEvent Modificado(Me)
        End Sub

        Public Sub Shading(ByVal IndiceRepresentacion As Integer, ByVal ColorShading As Color)
            If IndiceRepresentacion >= 0 AndAlso IndiceRepresentacion <= mRepresentaciones.GetUpperBound(0) Then
                mRepresentaciones(IndiceRepresentacion).Color = ColorShading
            End If
        End Sub

        Public Sub Reordenar()
            If Not Vacio Then
                Ordenamiento.Sort(Of ElementoZBuffer)(mObjetos)
            End If
        End Sub

        Public Function Pertenece(ByVal ParamArray Indices() As Double) As Integer
            If Not mObjetos Is Nothing Then
                For i As Integer = 0 To mObjetos.GetUpperBound(0)
                    If mObjetos(i).EsEquivalente(Indices) Then Return i
                Next

                Return -1
            Else
                Return -1
            End If
        End Function

        Public Function Pertenece(ByVal Indice As Integer, ByVal ValorIndice As Double) As Integer
            If Not mObjetos Is Nothing Then
                For i As Integer = 0 To mObjetos.GetUpperBound(0)
                    If Indice < mObjetos(i).NumeroIndices Then
                        If mObjetos(i).Indices(Indice) = ValorIndice Then Return i
                    End If
                Next

                Return -1
            Else
                Return -1
            End If
        End Function

        Public Sub New()
            MyBase.New()

            mTiempo = 0

        End Sub
    End Class
End Namespace

File: motor3d 2.0.ConvertedToC#/ZBuffer.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Xml.Linq;
using Motor3D.Primitivas3D;
using Motor3D.Utilidades;
using Motor3D.Espacio2D;
using System.Drawing;

namespace Motor3D.Escena
{
	public class ZBuffer : ObjetoEscena
	{

		public new event ModificadoEventHandler Modificado;
		public new delegate void ModificadoEventHandler(ref ZBuffer Sender);

		private ElementoZBuffer[] mObjetos;
		private Poligono2D[] mRepresentaciones;
		private int mNumeroIndices;

		private double mTiempo;
		public ElementoZBuffer[] Objetos {
			get { return mObjetos; }
		}

		public int NumeroObjetos {
			get {
				if ((mObjetos != null)) {
					return mObjetos.GetUpperBound(0) + 1;
				} else {
					return 0;
				}
			}
		}

		public Poligono2D[] Represenatciones {
			get { return mRepresentaciones; }
		}

		public bool Vacio {
			get { return mObjetos == null; }
		}

		public double TiempoUltimoCalculo {
			get { return mTiempo; }
		}

		public void Actualizar(ref Poliedro[] Poliedros, ref Camara3D Camara)
		{
			List<ElementoZBuffer> Objetos = new List<ElementoZBuffer>();
			List<Poligono2D> Representaciones = new List<Poligono2D>();
			System.DateTime T = DateAndTime.Now;
			int n = 0;
			if ((Poliedros != null)) {
				if (!Vacio) {
					for (int i = 0; i <= mObjetos.GetUpperBound(0); i++) {
						if (mObjetos[i].Indices[1] <= Poliedros.GetUpperBound(0) && mObjetos[i].Indices[2] < Poliedros[mObjetos[i].Indices[1]].NumeroCaras) {
							if (Poliedros[mObjetos[i].Indices[1]].CaraVisible[mObjetos[i].Indices[2], Camara]) {
								Objetos.Add(new ElementoZBuffer(Poliedros[mObjetos[i].Indices[1]].Caras[mObjetos[i].Indices[2]].BaricentroSRC.Z, n, mObjetos[i].Indices[1], mObjetos[i].Indices[2]));
								Representaciones.Add(new Poligono2D(Poliedros[mObjetos[i].Indices[1]].Caras[mObjetos[i].Indices[2]].Representacion[Poliedros[mObjetos[i].Indices[1]].Vertices]));
								Representaciones[Representaciones.Count - 1].Color = Poliedros[mObjetos[i].Indices[1]].Caras[mObjetos[i].Indices[2]].Color;
								Poliedros[mObjetos[i].Indices[1]].Caras[mObjetos[i].Indices[2]].CargadaEnBuffer = true;

								n += 1;
							} else {
								Poliedros[mObjetos[i].Indices[1]].Caras[mObjetos[i].Indices[2]].CargadaEnBuffer = false;
							}
						}
					}
				}

				for (int i = 0; i <= Poliedros.GetUpperBound(0); i++) {
					for (int j = 0; j <= Poliedros[i].NumeroCaras - 1; j++) {
						if (!Poliedros[i].Caras[j].CargadaEnBuffer) {
							if (Poliedros[i].CaraVisible[j, Camara]) {
								Objetos.Add(new ElementoZBuffer(Poliedros[i].Caras[j].BaricentroSRC.Z, n, i, j));
								Representaciones.Add(Poliedros[i].Representacion[j]);
								Poliedros[i].Caras[j].CargadaEnBuffer = true;

								n += 1;
							}
						}

					}
				}

				mObjetos = Objetos.ToArray();
				mRepresentaciones = Representaciones.ToArray();
				Reordenar();
			} else {
				mObjetos = null;
				mRepresentaciones = null;
			}

			mTiempo = (DateAndTime.Now - T).TotalMilliseconds;

			if (Modificado != null) {
				Modificado(this);
			}
		}

		public void Shading(int IndiceRepresentacion, Color ColorShading)
		{
			if (IndiceRepresentacion >= 0 && IndiceRepresentacion <= mRepresentaciones.GetUpperBound(0)) {
				mRepresentaciones[IndiceRepresentacion].Color = ColorShading;
			}
		}

		public void Reordenar()
		{
			if (!Vacio) {
				Ordenamiento.Sort<ElementoZBuffer>(ref mObjetos);
			}
		}

		public int Pertenece(params double[] Indices)
		{
			if ((mObjetos != null)) {
				for (int i = 0; i <= mObjetos.GetUpperBound(0); i++) {
					if (mObjetos[i].EsEquivalente(Indices))
						return i;
				}

				return -1;
			} else {
				return -1;
			}
		}

		public int Pertenece(int Indice, double ValorIndice)
		{
			if ((mObjetos != null)) {
				for (int i = 0; i <= mObjetos.GetUpperBound(0); i++) {
					if (Indice < mObjetos[i].NumeroIndices) {
						if (mObjetos[i].Indices[Indice] == ValorIndice)
							return i;
					}
				}

				return -1;
			} else {
				return -1;
			}
		}

		public ZBuffer() : base()
		{

			mTiempo = 0;

		}
	}
}

